[1단계 - 출석 구현] 루키(김준하) 미션 제출합니다.

junhaa opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [X] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [X] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
[프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/3972)
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [X] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->

* `OutputView` 출력 부분에서 메소드를 적절하게 분리하지 못하여 매개변수가 3개 이상인 경우가 있습니다.
* 모든 엔티티를 작게 분리하지 못하였습니다.



  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

안녕하세요 매트 😀 7기 BE 루키입니다.

사실 이번 출석부를 익숙하지 않은 TDD로 구현하면서, 시간이 부족했던 부분이 있었던 것 같아요. 이번 미션에서는 최대한 TDD에 적응해보고자 `RED -> GREEN -> Refactor` 사이클을 의식하면서 각 단계마다 커밋하다 보니 생각보다 시간이 부족했던 것 같아요.😭 구현에 최대한 중점을 두고 개발을 진행했더니 아직 코드에 다음과 같이 부족한 부분이 많이 있는 것 같아요.

1. 테스트 코드에 대한 리팩토링이 진행하지 못했어요. 테스트 코드에 대한 리팩터링은 최대한 미루자! 라고 하다보니 현재 테스트 코드가 매우 복잡한 상황이에요.😢
2. 도메인이 자신의 역할을 적절하게 가지고 있는 지 고민해보지 못했어요. 이 부분은 추후 바로 고민해보고 수정하려고 해요.

우선 이번 미션을 진행하면서 고민했던 부분이 몇 가지 있어요.

1. 테스트 검증 방법
     이번 미션에서는 TDD로 진행하면서 여러 테스트 코드를 작성해 보았는데요, 테스트 코드를 작성하는 와중에도 '내 테스트 코드가 신뢰할 수 있을까?' 라는 생각을 자주 했어요. 
     테스트를 검증하는 메서드에서 예외를 발생하는 경우는 예외를 검증하기 때문에 뚜렷하게 테스트 코드를 작성할 수 있었어요. 하지만 성공하는 테스트 코드를 어떤 방식으로 검증해야 하는지 궁금해요.
     
     
2. 출력 format
		이번 미션에서 출력하는 부분에 다음과 같은 내용들이 있었어요.
```
	오늘은 12월 13일 금요일입니다.
	12월 13일 금요일 09:59 (출석)
```

위와 같은 경우에는 `12월 13일 금요일` 이라는 내용이 중복되기 때문에 따로 메서드로 추출하였어요. 하지만 위 형식을 맞추기 위해 String 변환 format을 `"오늘은 %s 입니다."` 혹은 `"%s 09:59 (출석)"`과 같이 변경되어야 하는데, 이 부분에서 오히려 format에서 중복되는 부분을 추출하는 경우 format을 읽는 입장에서 가독성이 떨어지지 않을까? 라는 생각이 들었어요. `코드 중복 제거` vs `format에 대한 가독성` 중 어떤 부분이 맞는 지 고민되었어요.

saera-yook committed:
docs: README.md에 기능  목록 작성

saera-yook committed:
test: 출석, 지각, 결석 확인 실패하는 테스트 작성

saera-yook committed:
feat: 출석, 지각, 결석 확인 테스트 통과

saera-yook committed:
refactor: 출석 확인 메서드 이름 변경

saera-yook committed:
test: 월요일 출석, 지각, 결석 확인 실패하는 테스트 작성

saera-yook committed:
feat: 월요일 출석, 지각, 결석 확인 테스트 통과

saera-yook committed:
refactor: 불필요한 조건 삭제

saera-yook committed:
test: 캠퍼스 운영 시작 시간 확인 실패하는 테스트 작성

saera-yook committed:
feat: 캠퍼스 운영 시작 시간 확인 테스트 통과

saera-yook committed:
test: 캠퍼스 운영 종료 이후 출석 시 실패하는 테스트 작성

saera-yook committed:
feat: 캠퍼스 운영 종료 이후 출석 예외 테스트 통과

saera-yook committed:
test: 주말 및 공휴일 출석 가능 여부 확인 실패 테스트 작성

saera-yook committed:
feat: 주말 및 공휴일 출석 가능 여부 확인 테스트 통과

saera-yook committed:
test: 출석 정보 저장 실패 테스트 작성

saera-yook committed:
feat: 출석 정보 저장 테스트 통과

saera-yook committed:
refactor: AttendanceRepository 자료구조 변경

- 날짜와 시간 분리
- List<LocalDateTime> -> Map<LocalDate, HourMinute>
- HourMinute: 시간, 분 정보를 가진 레코드 클래스 생성

saera-yook committed:
refactor: domain 패키지 생성

saera-yook committed:
feat: 닉네임, 등교 시간 입력 받는 기능 구현

saera-yook committed:
refactor: Attendance에 name, timestamps 인스턴스 변수 추가

saera-yook committed:
test: 크루 출석 정보 저장 실패 테스트

saera-yook committed:
feat: 크루 출석 정보 저장 테스트 통과

saera-yook committed:
test: FileLoader의 Attendance 생성 실패 테스트 작성

saera-yook committed:
feat: FileLoader의 Attendance 생성 테스트 통과

saera-yook committed:
test: FileLoader의 Attendance 여러 개 생성 실패 테스트 작성

saera-yook committed:
feat: FileLoader의 Attendance 여러 개 생성 테스트 통과

saera-yook committed:
feat: 프로그램 시작 시 데이터를 attendances.csv에서 읽어오는 기능 구현

saera-yook committed:
test: 출석 정보 수정 성공테스트 실패하게 작성

saera-yook committed:
feat: 출석 정보 수정 기능 구현

saera-yook committed:
test: 출석 수정 시 기록이 이미 존재하는지 확인 테스트 작성

saera-yook committed:
feat: 출석 수정 시 이미 기록이 존재하는지 확인 기능 구현

saera-yook committed:
test: 특정 크루의 출석 기록 수정 테스트 작성

saera-yook committed:
feat: 특정 크루의 출석 기록 수정 기능 구현

saera-yook committed:
refactor: 캠퍼스 운영 규칙 관련 책임 분리

saera-yook committed:
feat: 출석, 지각, 결석 상태를 나타내는 Enum 생성

saera-yook committed:
refactor: AttendanceStatus Enum 적용

saera-yook committed:
test: 출석, 지각, 결석 횟수 조회 테스트 작성

saera-yook committed:
fix: 화~금 10시 이전 출석 시 결석 처리되는 오류 수정

saera-yook committed:
feat: 출석, 지각, 결석 횟수 조회 기능 구현

saera-yook committed:
fix: 결석 횟수 조회 오류 수정

saera-yook committed:
fix: 결석 횟수 조회 오류 수정

saera-yook committed:
test: 경고, 면담, 제적 확인 테스트 작성

saera-yook committed:
feat: 경고, 면담, 제적 확인 기능 구현

saera-yook committed:
test: 지각 3회를 결석 1회로 간주 기능 테스트 작성

saera-yook committed:
feat: 지각 3회를 결석 1회로 간주하는 기능 구현

saera-yook committed:
test: 크루 이름으로 제적 위험 여부 조회 테스트 작성

saera-yook committed:
feat: 크루 이름으로 제적 위험 여부 조회 기능 구현

saera-yook committed:
feat: 출석 등록 기능 구현

saera-yook committed:
feat: 출석 수정 기능 구현

saera-yook committed:
refactor: IOException -> RuntimeException 변환 처리

saera-yook committed:
fix: 메서드 분리에 따른 테스트 코드 수정

saera-yook committed:
feat: 크루별 출석 기록 확인 기능 구현

saera-yook committed:
feat: 제적 위험자 조회 기능 구현

saera-yook committed:
feat: 프로그램 기능 선택 및 실행하는 기능 구현

saera-yook committed:
refactor: OutputView 메서드 추출 및 상수화

saera-yook committed:
refactor: InputView 문자열 상수화

saera-yook committed:
refactor: HourMinute 시간 값이 없음을 의미하는 상수 추가

saera-yook committed:
refactor: DataFileReader 메서드 분리 및 indent 감소

saera-yook committed:
refactor: AttendanceController 불필요한 생성자 삭제

saera-yook committed:
refactor: Attendance 메서드 분리

saera-yook committed:
refactor: AttendanceRepository 불필요한 메서드 삭제

hyeonic commented on README.md:
```diff
@@ -1,3 +1,157 @@
 # java-attendance
 
-출석 미션 저장소
+## 기능 목록
```
요구사항을 잘 정돈 하셨네요 👍

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,68 @@
+        return new AttendanceRepository(FileLoader.loadAll(DataFileReader.read()));
+    }
+
+    private void initOperations(AttendanceRepository attendanceRepository) {
+        operations.put("1", () -> registerAttendance(attendanceRepository));
+        operations.put("2", () -> modifyAttendance(attendanceRepository));
+        operations.put("3", () -> queryAttendance(attendanceRepository));
+        operations.put("4", () -> queryWarningCrews(attendanceRepository));
+    }
```
아주 깔끔하게 분기 처리하였네요 ㅎㅎ `Runable`의 어떤 특성 때문에 람다로 표현이 가능할까요?!

junhaa replied:
```java
@FunctionalInterface
public interface Runnable {
    void run();
}
```
`Runnable` 인터페이스가 단 하나의 추상 메서드를 가지는 **함수형 인터페이스**라서 람다로 표현이 가능하다고 생각해요!
따라서 익명 클래스로 작성하는 것 보다 더욱 간결하게 작성할 수 있어 사용했습니다.

hyeonic replied:
추가로 사용하기에 따라 달라질 수 있지만 Runnable은 보통 `쓰레드풀`에 행위를 전달할 때 활용하곤 합니다. 크게 중요한 건 아니라 추후에 기회가 된다면 관련 키워드 학습해보시죠! (레벨 1에서는 다룰 내용은 아니라 자세한 설명은 스킵할게요)

junhaa replied:
Runnable과 쓰레드풀에 대해 조금 더 깊게 공부해야 겠네요 🤔 피드백 감사합니다 !

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,68 @@
+        AttendanceRepository attendanceRepository = initData();
+        initOperations(attendanceRepository);
+        String option;
+        while (!(option = getInputOption()).equals("Q")) {
+            operations.get(option).run();
+        }
```
`1, 2, 3, 4, Q` 등 각각의 커맨드가 명확히 어떤 목적인지 상수로 표현되면 더욱 좋을 것 같아요!

junhaa replied:
커맨드 상수로 분리하겠습니다!

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,68 @@
+
+    private AttendanceRepository initData(){
+        return new AttendanceRepository(FileLoader.loadAll(DataFileReader.read()));
+    }
```
컨벤션 한번 체크해주세요!
```suggestion
    private AttendanceRepository initData() {
        return new AttendanceRepository(FileLoader.loadAll(DataFileReader.read()));
    }
```

hyeonic commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,98 @@
+        return attendanceStatuses;
+    }
+
+    private void removeTodayStatus(final int today, final Map<AttendanceStatus, Integer> attendanceStatuses) {
+        LocalDate dateOfToday = LocalDate.of(2024, 12, today);
+        if(!timestamps.containsKey(dateOfToday)) {
+            return;
+        }
+        AttendanceStatus status = timestamps.get(dateOfToday).attendanceStatus();
+
+        attendanceStatuses.put(status, attendanceStatuses.get(status) - 1);
+    }
+
+    private void updateTimestampUntil(int today) {
+        for(int day = 1 ; day < today ; day++){
+            LocalDate date = LocalDate.of(2024, 12, day);
+            updateTimeStamp(day, date);
+        }
+    }
+
+    private void updateTimeStamp(final int day, final LocalDate date) {
+        if (AttendanceChecker.isCampusDay(day) && !timestamps.containsKey(date)){
+            timestamps.put(date, new HourMinute(NULL_TIME, NULL_TIME, AttendanceStatus.ABSENCE));
+        }
+    }
```
사소하지만 컨벤션 잘 확인해주세요. 다양한 사람들과 협업을 진행하게 되면 일관된 코드 컨벤션이 훨씬 읽기 수월하더라구요!

junhaa replied:
커밋전에 컨벤션을 확인하지 못한 부분들이 있네요.😭 전체적으로 적용해서 수정하겠습니다!

hyeonic commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    public Map<LocalDate, HourMinute> getTimestamps(int today) {
+        updateTimestampUntil(today);
+        return Collections.unmodifiableMap(timestamps);
+    }
```
`unmodifiableMap`을 활용하셨네요 👍 어떤 이점이 있을까요? 방어적 복사를 의도하신 걸까요?

junhaa replied:
현재 `getTimeStamps`메서드는 view에서 출력을 위해 getter를 사용하고 있는데요, 객체 변경을 방지하기 위해 `unmodifiableMap`을 사용했습니다!

hyeonic commented on src/main/java/attendance/domain/AttendanceChecker.java:
```diff
@@ -0,0 +1,74 @@
+import java.time.format.TextStyle;
+import java.util.Locale;
+
+public class AttendanceChecker {
```
AttendanceChecker에 속하는 대부분의 기능이 static 메서드를 통해 구성하셨는데요, 루키가 생각했을 때 객체지향 세계에서 static한 메서드의 사용의 장단점은 무엇이라 생각하시나요??

junhaa replied:
음.. 객체지향의 관점에서 static 사용에 대해 깊게 고민해본 적이 없었던 것 같아요.🤔

우선 제가 생각하는 static 메서드의 장점은 객체 를 생성하지 않고 호출할 수 있기 때문에 메모리를 절약 할 수 있다는 점이 있다고 생각했어요. 그래서 유틸과 같이 인스턴스 변수를 가지지 않는 특성을 가진 메서드를 static으로 선언하였어요.  

하지만 static 메서드가 객체를 생성하지 않기 때문에 오히려 객체지향의 관점에서 보면 단점도 많이 존재한다고 생각해요. static 메서드는 오버라이딩이 불가능 하기 때문에 상속과 다형성을 적용할 수 없다는 단점이 있어요.

결국 위와 같은 단점을 해결하기 위해 싱글턴 패턴을 적용한다면 객체를 생성해 상속과 다형성을 적용할 수 있을 것 같은데, 제가 생각하는 방식이 맞을까요..?!

hyeonic replied:
static한 메서드는 말 그대로 구현부를 유연하게 수정할 수 없다는게 가장 큰 단점인 것 같아요. 가령 예시로 들어 `checkRegularAttendance` 내의 조건이 수정되면 내부 수정의 영향으로 AttendanceChecker을 직접 의존하는 모든 객체에게 영향을 끼칠 수 있을 것 같아요.

하지만 AttendanceChecker를 인터페이스나 추상 클래스로 정의하면 어떻게 될까요? 추상화된 checkRegularAttendance 메서드는 구현체 객체의 특성에 따라 유연하게 로직을 추가할 수 있습니다. AttendanceChecker 인터페이스를 의존하는 객체는 단순히 새롭게 만든 구현체만 갈아끼면 되니까요 ㅎㅎ

junhaa replied:
추상화를 통해 static 메서드를 유연하게 수정할 수 있겠네요! `AttendanceChecker`에 해당하는 static 메서드를 추상화해서 변경하는 방법을 고민해봐야겠어요.🤔 구체적인 피드백 감사해요 매트 !

hyeonic commented on src/main/java/attendance/domain/AttendanceRepository.java:
```diff
@@ -0,0 +1,66 @@
+        Optional<Attendance> crewAttendance = attendances.stream()
+                .filter(attendance -> attendance.isNameMatch(name))
+                .findAny();
+        if (crewAttendance.isEmpty()) {
+            throw new IllegalArgumentException("[ERROR] 유효하지 않은 닉네임입니다.");
+        }
+        return crewAttendance.get();
```
Optional의 `orElseThrow` 통해 충분히 개선 가능해보이는데요! 관련 메서드를 활용하여 수정해보는 방향은 어떤가요?

junhaa replied:
`orElseThrow`를 이용해서 개선할 수 있을 것 같아요! 피드백 감사합니다😁

hyeonic commented on src/main/java/attendance/domain/WarningLevel.java:
```diff
@@ -0,0 +1,37 @@
+        this.level = level;
+    }
+
+    public static WarningLevel calculateLevel(final Map<AttendanceStatus, Integer> attendanceStatuses) {
+        int absenceCount = attendanceStatuses.get(AttendanceStatus.ABSENCE);
+        absenceCount += convertLateness(attendanceStatuses.get(AttendanceStatus.LATENESS));
+        if(absenceCount >= 6) {
+            return WarningLevel.REMOVE;
+        }
+        if(absenceCount >= 3) {
+            return WarningLevel.COUNSELING;
+        }
+        if(absenceCount >= 2) {
+            return WarningLevel.WARNING;
+        }
+        return WarningLevel.NONE;
+    }
```
10줄이 넘는 메서드인데요! 반복적되는 if문을 어떻게 수정할 수 있을까요?? 조금 더 힌트를 드리자면 루키가 구현하신 Controller의 operations 처럼 Functional Interface를 활용해봐도 좋을 것 같아요 ㅎㅎ

junhaa replied:
Functional Interface 사용해서 변경해보겠습니다!

hyeonic commented on src/main/java/attendance/domain/WarningLevel.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Map;
+
+public enum WarningLevel {
+    REMOVE("제적"), COUNSELING("면담"), WARNING("경고"), NONE("해당 없음");
+
+    private String level;
```
WarningLevel은 enum인데 level이 수정될 여지가 있을까요?? enum은 특성은 무엇일까요??

junhaa replied:
헉 `final` 키워드를 빼먹었네요.. 😮
제가 생각하는 Enum의 특성은 상수 값을 표현하는 클래스라고 생각합니다! 따라서 level 값 자체가 불변성을 가져야 한다고 생각해요.

hyeonic replied:
그 밖에도 equals, hashCode 등 일반적인 클래스랑 어떠한 차이가 있는지 알아봐도 좋을 것 같아요 ㅎㅎ

junhaa replied:
Enum의 경우에는 `equals`와 `hashCode`를 자동으로 재정의하는 특징이 있네요! 구체적으로 Enum에 대해서 더 공부해 봐야겠어요.🥲 답변 감사합니다!😊

hyeonic commented on src/main/java/attendance/util/FileLoader.java:
```diff
@@ -0,0 +1,36 @@
+        return attendance;
+    }
+
+    public static List<Attendance> loadAll(final List<String> datas) {
+        List<Attendance> attendances = new ArrayList<>();
+        datas.forEach(data -> {
+            List<String> seperatedData = Arrays.stream(data.split(",")).toList();
+            LocalDateTime localDateTime = LocalDateTime.parse(seperatedData.get(1).replace(" ", "T"));
+
+            Optional<Attendance> crewAttendance = attendances.stream()
+                    .filter(a -> a.isNameMatch(seperatedData.get(0)))
+                    .findFirst();
+
+            if (crewAttendance.isEmpty()) {
+                attendances.add(createAttendance(seperatedData.get(0), localDateTime));
+                return;
+            }
+            crewAttendance.get().add(localDateTime);
+        });
+        return attendances;
+    }
```
요 부분도 가능하면 depth 1로 줄여볼까요??

junhaa replied:
수정하도록 하겠습니다!

hyeonic commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.Map;
+
+public class OutputView {
+    private static final String DATE_FORMAT = "%02d월 %02d일 %s";
+    private static final String ATTENDANCE_RESULT_FORMAT = "%s %s (%s)\n";
+    private static final String MODIFY_SUCCESS_FORMAT = "\n%s %s (%s) -> %s (%s) 수정 완료!\n";
+    private static final String QUERY_ATTENDANCE_HEADER_FORMAT = "\n이번 달 %s의 출석 기록입니다.\n\n";
+    private static final String ATTENDANCE_STATUS_FORMAT = "%s: %d회\n";
+    private static final String ABSENCE_TIME_FORMAT = "--:--";
+    private static final String WARNING_FORMAT = "%s 대상자입니다.\n";
+    private static final String WARNING_CREW_HEADER_FORMAT = "제적 위험자 조회 결과\n";
+    private static final String WARNING_CREW_RESULT_FORMAT = "- %s: 결석 %d회, 지각 %d회 (%s)\n";
```
질문 주신 부분은 이런 format을 말씀하신 거겠죠?!

중복의 개수가 많지 않으면 의견 주신 것 처럼 반복되어도 크게 거슬리지 않을 것 같아요. 루키만의 혹은 팀 내에서 중복의 기준을 정하고 일관성만 맞춘다면 뭐든 상관 없다고 생각합니다!

hyeonic reviewed (changes requested):
안녕하세요 루키~ 출석 리뷰를 맡은 매트입니다 😃 꼼꼼한 요구사항 정의를 바탕으로 테스트까지 잘 작성하신 것 같아요! 고생하셨습니다 ㅎㅎ

간단히 질문 및 답변 남겼습니다! 이해하기 어렵거나 모호한 부분 있으면 언제든 추가 코멘트 및 DM 부탁드려요!

hyeonic commented on src/test/java/attendance/domain/AttendanceCheckerTest.java:
```diff
@@ -0,0 +1,118 @@
+    @Test
+    void test13() {
+        assertThatCode(() -> AttendanceChecker.validateCampusDay(24)).doesNotThrowAnyException();
+    }
```
> 1. 테스트 검증 방법

결국 루키가 검증하고자 하는 목적이 평일 출석인 경우 아무 일도 일어나지 않는 것이라면 충분히 납득할 수 있는 테스트 코드라고 생각해요! 저도 명확한 반환값이 없는 메서드의 성공 케이스는 종종 위와 같이 작성하곤 합니다.

junhaa committed:
refactor: Enum 상수에 해당하는 값 불변성 강화

- final 추가하여 상수에 대한 불변성 적용

junhaa committed:
style: 코드 컨벤션 적용

junhaa committed:
refactor: 문자열에 대한 접근 지정자 변경

- InputView에서만 사용하는 문자열 private으로 변경

junhaa committed:
refactor: 매직넘버 상수로 분리

- 사용자가 실행하고자 하는 기능의 번호를 상수로 분리

junhaa committed:
refactor: Optional 예외 처리 코드 변경

- orElseThrow()를 이용하여 코드 가독성 개선

junhaa committed:
refactor: WarningLevel을 계산하는 기능 분리

- WarningLevel 계산 기능을 WarningLevelCalculator 클래스에 두어 책임 분리
- WarningLevel을 계산하는 기능을 함수형 인터페이스로 구현하여 코드 길이 감소

junhaa committed:
refactor: 파일 데이터 불러오는 메서드 기능 분리

- 이름에 해당하는 Attendance 객체를 반환하는 기능으로 분리

junhaa committed:
fix: FileLoader에서 데이터가 추가되지 않는 오류 수정

junhaa committed:
refactor: 코드 가독성 개선

- BufferedReader의 lines()를 활용하여 코드 가독성 개선

junhaa committed:
fix: 예외 발생 시 프로그램을 종료하지 않도록 수정

- 잘못된 입력이 들어오는 경우 InputView에서 다시 입력받도록 변경

junhaa committed:
fix: 출석 기록 출력 시 불필요한 줄바꿈 제거

junhaa committed:
refactor: 테스트 코드 중복 제거 및 가독성 개선

- @ParameterizedTest를 이용하여 중복되는 코드 제거

junhaa committed:
style: 코드 컨벤션 적용

junhaa commented:
안녕하세요 매트😃
피드백 해주신 내용을 읽은 후 코드를 개선해 보았어요.

그리고 코드를 변경하면서 몇 가지 궁금한 점이 있었는데, 도움을 주셨으면 해요😭

## 1. 유틸리티 메서드의 분리

지난 리뷰의 피드백을 받고, static 메서드에 대해 고민해보았어요. 현재 AttendanceChecker와 WarningLevelCalculator 모두 특정 Enum을 반환하고 있는데, 이 경우 따로 클래스를 분리하는 것이 맞는 지, 아니면 Enum이 메서드를 관리해야 할 지 모르겠어요.

## 2. TDD 개발 방식

우선 이번 미션을 진행하면서, 설계에 대한 고민을 많이 하지 못해서 수정해야 할 부분들이 많이 보였어요. TDD방식을 진행하면서 우선 작은 기능부터 개발하면서 큰 틀을 생각하지 못한 것 같아요. 작은 기능부터 바로 개발하기 보다는 큰 틀을 먼저 생각하고 나서 기능을 분리한 후 TDD를 진행하는 것이 맞는 방법일까요?

수정해야 할 부분 피드백 남겨주시면 빨리 반영해볼게요!! 감사해요 매트😃

hyeonic commented on src/main/java/attendance/domain/HourMinute.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.time.LocalDateTime;
+
+public record HourMinute(int hour, int minute, AttendanceStatus attendanceStatus) {
+    public static final int NULL_TIME = -1;
+
+    public HourMinute(LocalDateTime localDateTime) {
+        this(localDateTime.getHour(), localDateTime.getMinute(), AttendanceChecker.checkAttendance(localDateTime));
+    }
+}
```
왜 java에서 기본제공하는 객체(e.g. LocalDateTime 등)를 사용하지 않고 별도 시간과 분을 다루는 객체를 구성하였을까요??

junhaa replied:
사실 처음에는 LocalDateTime과 같이 기본 객체를 사용하려고 했지만, 출석 기록이 존재하지 않는 경우에 대해서 어떻게 데이터를 저장해야 할까 고민하다가 새로운 객체로 구성했어요. 기본 객체를 사용할 수 있는 방법을 고민해봐야겠어요.

hyeonic commented on src/main/java/attendance/domain/WarningLevel.java:
```diff
@@ -0,0 +1,17 @@
+package attendance.domain;
+
+public enum WarningLevel {
+    REMOVE("제적"),
+    COUNSELING("면담"),
+    WARNING("경고"),
+    NONE("해당 없음");
+    private final String level;
+
+    WarningLevel(String level) {
+        this.level = level;
+    }
+
+    public String getLevel() {
+        return level;
+    }
+}
```
> 1. 유틸리티 메서드의 분리
지난 리뷰의 피드백을 받고, static 메서드에 대해 고민해보았어요. 현재 AttendanceChecker와 WarningLevelCalculator 모두 특정 Enum을 반환하고 있는데, 이 경우 따로 클래스를 분리하는 것이 맞는 지, 아니면 Enum이 메서드를 관리해야 할 지 모르겠어요.

`WarningLevelCalculator` 내에서 하는 행위를 WarningLevel 내부로 옮길 순 없을까요?? 좀 더 힌트를 드리면 enum 또한 필드로 `함수형 인터페이스`를 둘 수 있습니다! 다른 코드에서 사용하신 Functional Interface를 어떻게 적용할 수 있을지 고민해봐도 좋을 것 같아요 ㅎㅎ 가령 내부 필드로 Predicate를 두게 되면 true/false로 특정 조건(여기서는 absenceCount가 기준이 될 수 있겠네요)을 기반으로 WarningLevel을 응답할 수도 있겠네요!

junhaa replied:
Enum의 필드로 Predicate를 사용하는 방법은 생각을 못했는데, Enum 필드에 Functional Interface가 있으니 각각에 대해 더욱 쉽게 이해할 수 있는 것 같아요!! 좋은 피드백 정말 감사합니다 ☺️

hyeonic replied:
이러한 개념은 비단 enum에만 적용할 수 있는 것은 아니에요. 필드와 같이 값을 관리할 수 있는 것을 넘어 어떠한 `행위`를 필드로 가질 수 있고 이를 유연하게 확장할 수 있다는 점을 잘 알아두시면 좋을 것 같습니다 ㅎㅎ Java에서 `행위를 전달하고자 하는 함수`도 결국 `메서드를 하나만 가진 인터페이스` 일 뿐이니깐요

junhaa replied:
정말 기존에 값을 필드로 가지는 것만 생각하다가, 행위를 필드로 가진다는 것은 정말 좋은 방식인 것 같아요. 아직은 함수형 인터페이스를 활용하는 것이 어색하지만, 조금 더 익숙해져서 활용하게 된다면 다양한 방법으로 확장 할 수 있을 것 같아요. 피드백 정말 감사합니다 !!😊

hyeonic commented on src/main/java/attendance/util/FileLoader.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class FileLoader {
+
+    public static List<Attendance> loadAll(final List<String> datas) {
+        List<Attendance> attendances = new ArrayList<>();
+        datas.forEach(data -> {
+            List<String> seperatedData = Arrays.stream(data.split(",")).toList();
+            String name = seperatedData.get(0);
+            LocalDateTime localDateTime = LocalDateTime.parse(seperatedData.get(1).replace(" ", "T"));
+            findAttendanceByName(attendances, name).add(localDateTime);
+        });
+        return attendances;
+    }
+
+    private static Attendance findAttendanceByName(List<Attendance> attendances, String name) {
+        return attendances.stream()
+                .filter(attendance -> attendance.isNameMatch(name))
+                .findFirst()
+                .orElseGet(() -> {
+                    Attendance newAttendance = new Attendance(name);
+                    attendances.add(newAttendance);
+                    return newAttendance;
+                });
+    }
+}
```
FileLoader라는 클래스 이름을 가졌지만 File과 관련된 로직은 들어있지 않은 것 같아요. 객체가 하는 일을 잘 표현할 수 있는 이름으로 수정하는 방향은 어떨까요??

junhaa replied:
File에 대한 로직이 없는데 File로 클래스 이름을 지었네요. 조금 더 잘 표현할 수 있는 이름으로 수정해보겠습니다!😃

hyeonic commented on src/main/java/attendance/view/DataFileReader.java:
```diff
@@ -0,0 +1,29 @@
+public class DataFileReader {
+    private static final String FILE_PATH = "src/main/resources/attendances.csv";
+
+    public static List<String> read() {
+        List<String> datas = new ArrayList<>();
+        BufferedReader bufferedReader = createReader();
+
+        bufferedReader.lines()
+                .skip(1)
+                .forEach(data -> datas.add(data));
+        return datas;
+    }
```
`BufferedReader`를 close 시키지 않으면 어떤 문제가 발생할 수 있을까요? `Closable`, `try-with-resources` 관련 키워드 통해 추가 학습해보면 좋을 것 같아요 ㅎㅎ

junhaa replied:
`BufferedReader`를 `close()`하지 않는 경우에는 GC가 소멸시킬 때 까지 남아있어 메모리 누수가 발생할 수 있다고 생각해요. `try-with-resources`방식으로 close하도록 변경하겠습니다!

hyeonic commented on src/test/java/attendance/domain/AttendanceCheckerTest.java:
```diff
@@ -0,0 +1,92 @@
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+@DisplayName("출석 상태를 판별")
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
```
오호 이런 애노테이션이 있었군요 ㅎㅎ 👍

hyeonic commented on src/main/java/attendance/domain/WarningLevelCalculator.java:
```diff
@@ -0,0 +1,51 @@
+    private static final int NONE_ABSENCE_BOUND = 0;
+
+    static {
+        CALCULATE_LEVEL = new LinkedHashMap<>();
```
`LinkedHashMap`은 어떠한 목적으로 사용하셨나요? 내부적으로 어떤 특징을 가지고 있을까요?

junhaa replied:
`LinkedHashMap`은 `key`값을 저장 순서대로 보장해주는 특징이 있어요. 이러한 특징을 이용해 순차적으로 Predicate를 검증하기 위해 사용했어요!

hyeonic commented on src/main/java/attendance/domain/AttendanceChecker.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    public static boolean isCampusDay(final int day) {
+        LocalDate localDate = LocalDate.of(2024, 12, day);
+        DayOfWeek dayOfWeek = localDate.getDayOfWeek();
+        return !(isHoliday(day) || isWeekend(dayOfWeek));
+    }
+
+    private static boolean isHoliday(final int day) {
+        return day == 25;
+    }
```
일부러 12월로 한정지은 걸까요??

junhaa replied:
사실 이 부분을 많이 고민했었어요. 현재 시간을 기준으로 기능을 구현하는 것이 맞다고 생각했지만, 우선 .csv 파일의 데이터를 기존의 데이터로 사용하다 보니 12월의 데이터를 사용하게 되었어요😭 그리고 무엇보다 가장 고민이 되었던 점이 **'출석 기록의 시작을 언제부터로 정해야 하는가'** 였어요. 이러한 문제 때문에 많은 고민을 하다가, 우선 12월 한 달을 기준으로 기능을 구현하게 된 것 같아요.

hyeonic replied:
보통 한달을 단위로 출석을 집계하니 요구사항을 좀 더 다양한 케이스로 바라보는 것도 좋을 것 같네요! 결국 우리가 개발하는 애플리케이션은 지속해서 발전해가야 하는데 이렇게 타이트한 비즈니스 로직이 추후 발목을 잡기도 하더라구요 ㅜㅠ

junhaa replied:
사실 매트가 피드백 해주신 것 처럼 실제 비즈니스에서 활용할 수 없다는 점이 크게 다가와서 코드 수정을 시도했었어요.😅 하지만 현재 코드 내부에서 고쳐야 할 부분도 많고, 출석 기록의 기준을 정하지 못해서 고치다가 롤백을 했어요...🥲 다음 step2에서는 한달을 단위로 현재 날짜를 기준으로 출석하는 애플리케이션을 구현하도록 노력해볼게요 !!

hyeonic commented on src/main/java/attendance/domain/AttendanceChecker.java:
```diff
@@ -0,0 +1,74 @@
+        return checkRegularAttendance(hour, minute);
+    }
+
+    public static AttendanceStatus checkRegularAttendance(final int hour, final int minute) {
+        if ((hour == 10 && minute <= 5) || hour < 10) {
+            return PRESENT;
+        }
+        if (hour == 10 && (minute <= 30)) {
+            return LATENESS;
+        }
+        return ABSENCE;
+    }
+
+    public static AttendanceStatus checkAttendance(final int hour, final int minute) {
+        if (hour <= 13 && minute <= 5) {
+            return PRESENT;
+        }
+        if (hour == 13 && (minute <= 30)) {
+            return LATENESS;
+        }
+        return ABSENCE;
+    }
```
관련 로직을 AttendanceStatus 내로 위임하는 방법은 어떤가요?

junhaa replied:
`AttendanceStatus`를 반환하는 로직이라 `AttendanceStatus` 내로 위임하는 것이 맞는 것 같아요. 피드백 감사합니다!

hyeonic reviewed (changes requested):
안녕하세요 루키! 답변 주신 부분 확인하였습니다. 추가 코멘트 남겼으니 확인 부탁드려요 ㅎㅎ

모호하거나 이해 안가는 부분 있다면 편하게 DM 남겨주세요!

hyeonic commented:
> 2. TDD 개발 방식
우선 이번 미션을 진행하면서, 설계에 대한 고민을 많이 하지 못해서 수정해야 할 부분들이 많이 보였어요. TDD방식을 진행하면서 우선 작은 기능부터 개발하면서 큰 틀을 생각하지 못한 것 같아요. 작은 기능부터 바로 개발하기 보다는 큰 틀을 먼저 생각하고 나서 기능을 분리한 후 TDD를 진행하는 것이 맞는 방법일까요?

취향에 따라 다를 수 있지만 최대한 작은 단위로 쪼개야 TDD하기 용이하더라구요. 검증의 범위가 좁아져야 빠르게 테스트 피드백을 받아볼 수도 있구요. 아직 익숙하지 않아 어려운 과정이라고 생각이 되어요. 그렇기 때문에 루키가 잘 정돈하신 것 처럼 전체적인 요구사항을 기반으로 체크 리스트를 작은 단위로 쪼개는 연습을 하다보면 자연스럽게 검증하고자 하는 범위가 좁아지지 않을까 생각이 듭니다 ㅎㅎ

junhaa committed:
refactor: 객체의 역할에 맞게 클래스 이름 수정

- 데이터로 불러오는 기능에 맞게 `FileLoader` -> `DataLoader`로 변경

junhaa committed:
refactor: AttendanceStatus를 반환하는 메서드 AttendanceStatus로 위임

junhaa committed:
refactor: try-with-resources를 사용하여 자동 close하도록 변경

junhaa committed:
refactor: WarningLevelCalculator의 메서드를 WarningLevel로 위임

- WarningLevel을 반환하는 메서드를 WarningLevel에서 관리하도록 변경
- Predicate를 Enum의 필드로 두어 WarningLevel 반환 검증에 사용

junhaa committed:
fix: 종료 기능 입력 시 예외가 발생하는 오류 수정

junhaa committed:
refactor: 매개변수 명 변경

junhaa committed:
refactor: 목적에 맞게 클래스 명 변경

- 크루들의 데이터를 저장하는 의미에 맞게AttendanceRepository -> Crews로 변경

junhaa committed:
refactor: 출석 데이터 클래스 분리

기존 Attendance 클래스에서 가지는 Map을 클래스로 분리

junhaa committed:
refactor: 목적에 맞게 클래스 명 변경

- 크루의 이름과 출석 데이터를 저장하는 클래스의 이름을 Attendance -> Crew 로 목적에 맞게 변경

junhaa committed:
fix: 출석 수정 시 출석 기록이 없는 경우 예외가 발생하지 않는 오류 수정

junhaa committed:
refactor: static 메서드를 인터페이스로 추상화

- 인터페이스로 추상화 하여 static을 사용하지 않고 객체를 생성하도록 변경

junhaa committed:
style: 코드 컨벤션 적용

junhaa commented:
안녕하세요 매트 😃
매트가 해주신 피드백을 바탕으로 코드를 수정해봤어요.

- 전체적인 클래스 분리 및 이름 수정
- 메서드 로직 클래스에 맞게 위임
- 예외 발생 시 재입력하는 기능 수정
- `BufferedReader` 객체 `close`
- `AttendanceChecker` 추상화를 통해 static 메서드를 변경

이번 코드를 수정하면서 다음과 같은 궁금한 점이 생겼어요.

### 1. 객체의 체이닝 

현재 변경한 코드는 출석 기능에서 다음과 같이 객체에 요청하는 방식을 사용하고 있어요.

`Crews` -> `Crew` -> `Attendances` -> `Attendance.attend()`

실제 출석 기능을 처리하는 메서드는 `Attendance`인데, `Attendances`에서는 단지 `Attendance.attend()` 메서드를 호출하는 메서드만 만드는 것이 맞는지 궁금해요. 그리고 만약에 이런 식으로 객체에 대한 호출이 많아진다면, 어떻게 체이닝이 깊어지지 않도록 끊어내야 할 지 모르겠어요.

### 2. `AttendanceChecker` 구현체의 위치

이전에 매트가 남겨주신 피드백에서 `AttendanceChecker`를 추상화 하는 방법에 대해 고민해보았어요.🤔 인터페이스를 이용하여 추상화 하였는데(사실 이 방법도 맞는 지 모르겠어요😭), 인터페이스의 구현체를 어디서 생성하고, 어떤 클래스가 가지고 있어야 할 지 모르겠어요.

이번에도 수정해야 할 부분이 많아보이지만, 피드백 남겨주시면 최대한 빨리 반영해볼게요!😃

hyeonic commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,51 @@
+        return checkRegularAttendance(hour, minute);
+    }
+
+    private static AttendanceStatus checkMondayAttendance(final int hour, final int minute) {
+        if (hour <= 13 && minute <= 5) {
+            return PRESENT;
+        }
+        if (hour == 13 && (minute <= 30)) {
+            return LATENESS;
+        }
+        return ABSENCE;
+    }
+
+    private static AttendanceStatus checkRegularAttendance(final int hour, final int minute) {
+        if ((hour == 10 && minute <= 5) || hour < 10) {
+            return PRESENT;
+        }
+        if (hour == 10 && (minute <= 30)) {
+            return LATENESS;
+        }
+        return ABSENCE;
+    }
```
이런 부분들도 때에 따라 Predicate 활용하여 enum 내에 각각의 인스턴스가 행위를 가지고 있을 수 있겠죠??

junhaa replied:
우와.. 그렇네요..😮 조건문에 해당하는 부분을 `Predicate`를 이용해서 변환하는 방법도 많이 고민해봐야겠어요. 피드백 감사합니다!

hyeonic commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Attendances {
+    private final Map<LocalDate, Attendance> attendances;
+    private final AttendanceChecker checker;
+
+    public Attendances() {
+        this.attendances = new HashMap<>();
+        this.checker = new RegularAttendanceChecker();
+    }
```
AttendanceChecker를 인터페이스로 잘 구현하셨지만 결국 직접적인 구현체를 의존하는 형태가 된 것 같아요. `생성자 주입`과 `다형성`에 대해 학습하면 좋을 것 같습니다 ㅎㅎ

junhaa replied:
이 부분은 아직 부족한 부분이 있어서 학습한 후 조금 더 고민해봐야겠어요.🫠

hyeonic commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,88 @@
+        return Collections.unmodifiableMap(attendances);
+    }
+
+    public Map<AttendanceStatus, Integer> countAttendanceStatus(final int today) {
+        updateTimestampUntil(today);
+
+        Map<AttendanceStatus, Integer> attendanceStatuses = calculateAttendanceStatus(today);
+
+        return attendanceStatuses;
+    }
```
바로 리턴 가능할 것 같아요!

```suggestion
    public Map<AttendanceStatus, Integer> countAttendanceStatus(final int today) {
        updateTimestampUntil(today);

        return calculateAttendanceStatus(today);
    }

```

hyeonic commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,88 @@
+        return attendances.containsKey(localDate);
+    }
+
+    private void fillEmptyAttendance(final int day, final LocalDate date) {
+        if (checker.isCampusDay(day) && !attendances.containsKey(date)) {
+            attendances.put(date, new Attendance(null, AttendanceStatus.ABSENCE));
+        }
+    }
```
파라미터로 전달할 때 final 유무가 일관성이 없는 것 같아요. 하나로 통일하는 방향은 어떠신가요?

junhaa replied:
Intellij에서 자동으로 메서드를 생성하는 경우 자동으로 final이 추가되는 경우가 있어 코드마다 final이 일관되지 않게 추가되어 있었어요.  PR 이전에 한번 정리를 해야 겠다고 생각했는데 까먹고 수정을 못했네요 ㅠ🥲

hyeonic commented on src/main/java/attendance/domain/RegularAttendanceChecker.java:
```diff
@@ -0,0 +1,43 @@
+import java.time.format.TextStyle;
+import java.util.Locale;
+
+public class RegularAttendanceChecker implements AttendanceChecker {
+
+    @Override
+    public void checkCampusHour(final int hour, final int minute) {
+        if (hour < 8 || (hour >= 23 && minute > 0)) {
+            throw new IllegalArgumentException("[ERROR] 현재 캠퍼스 운영시간이 아닙니다.");
+        }
+    }
+
+    @Override
+    public void validateCampusDay(final int day) {
+        LocalDate localDate = LocalDate.of(2024, 12, day);
+        DayOfWeek dayOfWeek = localDate.getDayOfWeek();
+        if (!isCampusDay(day)) {
+            throw new IllegalArgumentException(
+                    String.format("[ERROR] %d월 %d일 %s은 등교일이 아닙니다.", localDate.getMonthValue(), day,
+                            dayOfWeek.getDisplayName(
+                                    TextStyle.FULL, Locale.KOREA)));
+        }
+    }
+
+    @Override
+    public boolean isCampusDay(final int day) {
+        LocalDate localDate = LocalDate.of(2024, 12, day);
+        DayOfWeek dayOfWeek = localDate.getDayOfWeek();
+        return !(isHoliday(day) || isWeekend(dayOfWeek));
+    }
+
+    private static boolean isHoliday(final int day) {
+        return day == 25;
+    }
+
+    private static boolean isWeekend(final DayOfWeek dayOfWeek) {
+        return dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY;
+    }
+}
```
> 2. AttendanceChecker 구현체의 위치
이전에 매트가 남겨주신 피드백에서 AttendanceChecker를 추상화 하는 방법에 대해 고민해보았어요.🤔 인터페이스를 이용하여 추상화 하였는데(사실 이 방법도 맞는 지 모르겠어요😭), 인터페이스의 구현체를 어디서 생성하고, 어떤 클래스가 가지고 있어야 할 지 모르겠어요.

이것은 전적으로 어떠한 목적으로 인터페이스로 추상화 했는지에 대해서 집중해보면 좋을 것 같아요. 만약 구현체에 도메인에 의존하면 안되는 부분이 존재(csv를 읽는 것 등)한다고 가정하면 도메인 내에서는 인터페이스만을 의존하여도 로직을 수행할 수 있도록 구현하고 도메인 영역 진입 시점에 생성자 주입 혹은 메서드를 통해 주입하는 시점에 실제 구현체를 넣게 되면 도메인 영역 내에서는 구체적인 의존성을 피할 수 있습니다.

한 가지 예시를 더 들어볼게요. 출석 애플리케이션에서는 csv라는 파일을 통해 출석 정보를 관리하고 있죠? 하지만 file을 읽고 이를 적절히 추출하는 것은 너무 구체적이며 입출력 형식에 따라 언제든 바뀔 수 있기 때문에 도메인 영역에 해당하지 않는다고 생각해요. 하지만 추후에 csv가 아닌 다른 파일 형식으로 요구사항이 변경되면 어떻게 될까요?? 좀 더 큰 관점으로 봐서 `Crews`를 저장하고 조회하기 위한 어떠한 인터페이스를 추상화하고 요구사항에 맞춰 구현체를 다시 구현하여 활용할 수도 있구요!

hyeonic reviewed (approved):
안녕하세요 루키! 처음 리뷰 요청했을 때 보다 훨씬 정돈된 것 같아요 ㅎㅎ 고생하셨습니다 😃

먼저 코드 내에 남기지 못한 질문 부터 답변 드릴게요.

> 1. 객체의 체이닝

그렇다면 무의미하게 객체를 포장한 것이 아닌지 의심해볼 수도 있을 것 같아요. 하지만 왜 두 객체를 하나로 묶어서 다뤄야만 하는지 명확한 이유만 있다면 반복된 체이닝도 자연스럽게 근거가 생길 수 있다고 생각합니다.

끊어내는 것은 결국 해당 객체가 가진 책임을 적절히 위임하는 방법일 것 같아요. getter를 최대한 지양하고 얻고자하는 것을 위해 메시지를 던지는 형태로요!

---

2단계는 1단계 + 리뷰를 기반으로 TDD를 적극 활용하여 `출석 다시 구현하기`를 하는 것으로 알고 있는데요! 출석 도메인에 대한 이해도가 많이 올라간 것 같아서 이번에는 적극적으로 TDD를 도입하면서 부족했던 단위 테스트나 고민했던 포인트들도 잘 적용해보면 좋을 것 같아요!

간단히 코멘트 남겼습니다! 1단계는 이만 마무리하고 2단계에서 추가 고려해주세요 ㅎㅎ 다음 단계도 잘 부탁드립니다 😃

긴가민가한 부분 있다면 언제든 DM 주세요!
